File: lib/ts/src/internal/ProcessBytes.ts
//|
//|  Copyright The Telepact Authors
//|
//|  Licensed under the Apache License, Version 2.0 (the "License");
//|  you may not use this file except in compliance with the License.
//|  You may obtain a copy of the License at
//|
//|  https://www.apache.org/licenses/LICENSE-2.0
//|
//|  Unless required by applicable law or agreed to in writing, software
//|  distributed under the License is distributed on an "AS IS" BASIS,
//|  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//|  See the License for the specific language governing permissions and
//|  limitations under the License.
//|

import { Serializer } from '../Serializer.js';
import { TelepactSchema } from '../TelepactSchema.js';
import { Message } from '../Message.js';
import { handleMessage } from '../internal/HandleMessage.js';
import { parseRequestMessage } from '../internal/ParseRequestMessage.js';
import { Response } from '../Response.js';
import { SerializationError } from '../SerializationError.js';
import { TelepactError } from '../TelepactError.js';
import { UpdateHeaders } from '../Server.js';
import { buildUnknownErrorMessage } from './UnknownError.js';

export type ErrorHandler = (error: TelepactError) => void;
export type RequestHandler = (message: Message) => void;
export type ResponseHandler = (message: Message) => void;
export type AuthHandler = (headers: Record<string, any>) => Record<string, any> | Promise<Record<string, any>>;
export type FunctionRouter = {
    route: (message: Message) => Promise<Message>;
    requiresAuthentication: (functionName: string) => boolean;
};
export type Middleware = (requestMessage: Message, functionRouter: FunctionRouter) => Promise<Message>;

export async function processBytes(
    requestMessageBytes: Uint8Array,
    updateHeaders: UpdateHeaders | undefined,
    serializer: Serializer,
    telepactSchema: TelepactSchema,
    onError: ErrorHandler,
    onRequest: RequestHandler,
    onResponse: ResponseHandler,
    onAuth: AuthHandler,
    middleware: Middleware,
    functionRouter: FunctionRouter,
): Promise<Response> {
    try {
        const requestMessage = parseRequestMessage(requestMessageBytes, serializer, telepactSchema, onError);

        try {
            onRequest(requestMessage);
        } catch (error) {
            // Handle error
        }

        const responseMessage = await handleMessage(
            requestMessage,
            updateHeaders,
            telepactSchema,
            middleware,
            functionRouter,
            onError,
            onAuth,
        );

        try {
            onResponse(responseMessage);
        } catch (error) {
            // Handle error
        }

        let responseBytes: Uint8Array;
        try {
            responseBytes = serializer.serialize(responseMessage);
        } catch (error) {
            const wrapped = error instanceof SerializationError
                ? new TelepactError('telepact response serialization failed', 'serialization', error)
                : new TelepactError('telepact server processing failed while serializing the response', 'serialization', error);
            try {
                onError(wrapped);
            } catch (error) {
                // Handle error
            }
            const unknownResponseBytes = serializer.serialize(buildUnknownErrorMessage(wrapped));
            return { bytes: unknownResponseBytes, headers: {} };
        }

        return { bytes: responseBytes, headers: responseMessage.headers };
    } catch (error) {
        const wrapped = error instanceof TelepactError
            ? error
            : error instanceof SerializationError
                ? new TelepactError('telepact response serialization failed', 'serialization', error)
                : new TelepactError('telepact server processing failed', undefined, error);
        try {
            onError(wrapped);
        } catch (error) {
            // Handle error
        }

        const responseBytes = serializer.serialize(buildUnknownErrorMessage(wrapped));

        return { bytes: responseBytes, headers: {} };
    }
}

File: lib/go/internal/ProcessBytes.go
//|
//|  Copyright The Telepact Authors
//|
//|  Licensed under the Apache License, Version 2.0 (the "License");
//|  you may not use this file except in compliance with the License.
//|  You may obtain a copy of the License at
//|
//|  https://www.apache.org/licenses/LICENSE-2.0
//|
//|  Unless required by applicable law or agreed to in writing, software
//|  distributed under the License is distributed on an "AS IS" BASIS,
//|  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//|  See the License for the specific language governing permissions and
//|  limitations under the License.
//|

package internal

import "fmt"

type FunctionRouter interface {
	Route(ServerMessage) (ServerMessage, error)
	RequiresAuthentication(functionName string) bool
}

type Middleware func(ServerMessage, FunctionRouter) (ServerMessage, error)

func ProcessBytes(
	requestMessageBytes []byte,
	updateHeaders func(map[string]any),
	deserialize func([]byte) (ServerMessage, error),
	serialize func(ServerMessage) ([]byte, error),
	schema SchemaAccessor,
	onError func(error),
	onRequest func(ServerMessage),
	onResponse func(ServerMessage),
	onAuth func(map[string]any) <-chan map[string]any,
	middleware Middleware,
	functionRouter FunctionRouter,
) (ServerMessage, []byte, error) {
	requestMessage, err := ParseRequestMessage(requestMessageBytes, deserialize, schema, onError)
	if err != nil {
		return ServerMessage{}, nil, err
	}

	safeInvokeMessage(onRequest, requestMessage)

	responseMessage, err := HandleMessage(requestMessage, updateHeaders, schema, middleware, functionRouter, onError, onAuth)
	if err != nil {
		wrapped := wrapUnknownError(err, "telepact server processing failed", "")
		invokeOnError(onError, wrapped)
		return buildUnknownResponse(serialize, wrapped.CaseID())
	}

	safeInvokeMessage(onResponse, responseMessage)

	responseBytes, err := serialize(responseMessage)
	if err != nil {
		wrapped := wrapUnknownError(
			fmt.Errorf("telepact response serialization failed: %w", err),
			"telepact response serialization failed",
			"serialization",
		)
		invokeOnError(onError, wrapped)
		return buildUnknownResponse(serialize, wrapped.CaseID())
	}

	return responseMessage, responseBytes, nil
}

func safeInvokeMessage(callback func(ServerMessage), message ServerMessage) {
	if callback == nil {
		return
	}
	defer func() { _ = recover() }()
	callback(message)
}

func buildUnknownResponse(serialize func(ServerMessage) ([]byte, error), caseID string) (ServerMessage, []byte, error) {
	unknownMessage := newUnknownErrorResponse(map[string]any{}, caseID)

	responseBytes, err := serialize(unknownMessage)
	if err != nil {
		return ServerMessage{}, nil, err
	}

	return unknownMessage, responseBytes, nil
}
